Progress Report - October 2022 (#336)
yuzu-emu/yuzu-emu.github.io: site/content/entry/yuzu-progress-report-oct-2022/index.md

byte[] investigated the log file, and stubbed the previously unimplemented calls {{< gh-hovercard "9032" "CheckFriendListAvailability" >}} and {{< gh-hovercard "9033" "GetCacheStorageSize," >}} allowing `No Man's Sky` to boot without any stubbing.

{{< imgs
    "./nms.png| Still the best screensaver"
  >}}

`Bayonetta 3`'s release caused some issues that were tricky to track down. 
Your resident [bunnei](https://github.com/bunnei) rabbit properly implemented `ListOpenContextStoredUsers`, and the related `StoreOpenContext` which is used to set stored opened user profiles.
`ListOpenContextStoredUsers` was causing `Bayonetta 3` to crash on startup. 
{{< gh-hovercard "9157" "Fixing the stubs" >}} allowed the game to finally boot, and while it is now playable, it is quite slow still. 
Stay tuned for performance optimizations regarding this game!

[german77](https://github.com/german77) took part in the fun, {{< gh-hovercard "9149" "stubbing the SetRecordVolumeMuted service," >}} allowing the game to boot.

{{< imgs
    "./bayo.png| Hot protagonists in your area (Bayonetta 3)"
  >}}

> Another notable release from bunnei this month was the beginning of the {{< gh-hovercard "9071" "multi-process project." >}} 
> Since the very beginning, yuzu has only ever supported a single-process architecture, since games only ever would use one process on the Switch. 
> However, to help improve yuzu's accuracy to the Switch software, yuzu has been focusing on getting key operating system features working, and one of the most important is multi-process support, since this corresponds to how the Switch software actually works.

Reviewer: ```suggestion
Another notable release from bunnei this month was the start of the {{< gh-hovercard "9071" "multiprocess project." >}} 
Since the very beginning, yuzu has only ever supported a single-process architecture, as games only ever use one process on the Switch. 
However, to help improve yuzu's accuracy for Switch software, yuzu has been focusing on getting key operating system features working, and one of the most important is multiprocess support, since this corresponds to how Switch software actually works.
```

> byte[] and bunnei have been hard at work this month getting certain prerequisites available for multi-process support. One of those is {{< gh-hovercard "9055" "support for server sessions," >}}  needed for the nx-hbloader homebrew. 
> This is used on the Switch to launch homebrew, and is needed for homebrew that do not work when launched directly from their NRO file.
> 
> Additionally, byte[] also needed to {{< gh-hovercard "9137" "fix several services," >}} as the previous stub implementations had assumed that they would never be torn down. 
> With these changes, and the associated graphical fixes, nx-hbloader and nx-hbmenu now finally work as intended! 
> To use them, you can extract hbl.nsp and hbmenu.nro from your Atmosphere installation. 
> Place hbmenu.nro in yuzu's SD card directory, and then launch hbl.nsp, and the homebrew menu will boot:

Reviewer: ```suggestion
byte[] and bunnei have been hard at work getting certain prerequisites available for multiprocess support. One of those is {{< gh-hovercard "9055" "support for server sessions," >}}  needed for the `nx-hbloader` homebrew. 
This is used on the Switch to launch homebrew, and is needed for homebrew that do not work when launched directly from their NRO file.

Additionally, byte[] needed to {{< gh-hovercard "9137" "fix several services," >}} as the previous stub implementations had assumed that they would never be torn down. 
With these changes, and the associated graphical fixes, `nx-hbloader` and `nx-hbmenu` now finally work as intended! 
To use them, you can extract `hbl.nsp` and `hbmenu.nro` from your Atmosphere installation. 
Place `hbmenu.nro` in yuzu's SD card directory, and then launch `hbl.nsp`, and the homebrew menu will boot.
```

{{< imgs

>     "./hbmenu.mp4| They launch now"

Reviewer: ```suggestion
    "./hbmenu.mp4| They launch now (Homebrew menu and Sonic 1 2013)"
```

  >}}

> Let’s talk about amiibos.
> Amiibo data is stored both encrypted and as plain data. 
> To access the encrypted portion, a key dump is required, but not all games use the encrypted portion, some are just fine with the plain text available.
> For those cases, german77 now {{< gh-hovercard "9113" "marks amiibos as read-only if no key dumps are available." >}}
> This won’t replace dumping your amiibo keys if a game needs it, but many games will be fine with just a basic level of read-only detection.

Reviewer: ```suggestion
Let’s talk about amiibos.
Amiibo data is stored both encrypted and as plain data. 
To access the encrypted portion, a key dump is required, but not all games use the encrypted portion and are just fine with the plaintext available.
For those cases, german77 now {{< gh-hovercard "9113" "marks amiibos as read-only if no key dumps are available." >}}
This won’t replace dumping your amiibo keys if a game needs it, but many games will be fine with just a basic level of read-only detection.
```

{{< single-title-imgs
    "Thank you Gidoly and german77 for the pics!"
    "./amiibo2.png"
    "./amiibo3.png"
    "./amiibo1.png"
    >}}

## Input improvements

> Here’s some Switch Vs PC compatibility debate. 
> The emulator performs vibration tests regularly each time the game initializes a controller for vibration, the problem (there’s always a problem) is that some controllers take longer than others to respond.
> Previously, a fixed 15ms delay was implemented, because some controllers just needed that much time to respond.
> If the game performs successive vibration tests, the game is delayed until the controller responds, reducing framerates, for example, from 60 to 32 FPS. 
> All that performance being lost waiting for a bad quality controller.

Reviewer: ```suggestion
yuzu performs vibration tests every time a game initializes a controller for vibration, but the problem (there’s always a problem) is that some controllers take longer than others to respond.
Previously, a fixed 15ms delay was implemented, because some controllers simply needed that much time to respond.
If the game performs successive vibration tests, the game is delayed until the controller responds, which results in a reduced framerate, for example, from 60 to 32 FPS. 
All that performance being lost waiting on a bad quality controller is a tragedy.
```

> So, a better middle ground is needed. Thankfully german77 came with a much smarter solution.
> Instead of repeating the test each time, it’s better to {{< gh-hovercard "9107" "cache it the first time and return that same result afterwards." >}}
> A nice free performance boost for generic controller users.
> 
> Newcomer [ZwipZwapZapony](https://github.com/ZwipZwapZapony) (love the name) {{< gh-hovercard "9008" "fixed a copy-paste oopsie in how left and right controller colours are identified." >}}
> Thank you!

Reviewer: ```suggestion
So, a better middle ground is needed. Thankfully, german77 came up with a much smarter solution.
Instead of repeating the test each time, it’s better to {{< gh-hovercard "9107" "cache it the first time and return that same result afterwards." >}}
A nice, free, performance boost for those of us who use generic controllers.

Newcomer [ZwipZwapZapony](https://github.com/ZwipZwapZapony) (love the name) {{< gh-hovercard "9008" "fixed a copy-paste oopsie in how left and right controller colours are identified." >}}
Thank you!
```

## Audio changes

> Ever noticed how in motherboard BIOS/UEFI settings, the Auto option seems to do nothing, instead just being the first option available?
> Well, we cheated you exactly like that for years, the “auto” audio backed option only selected cubeb every time by default. The shame.
> Maide found out that some audio devices incur a huge latency when using cubeb, so he decided to {{< gh-hovercard "9039" "perform a latency test" >}} and select either cubeb or SDL based on the results.

Reviewer: ```suggestion
Have you ever noticed how in motherboard BIOS/UEFI settings, the `Auto` option seems to do nothing, and instead simply reflects the first option available in the list?
Well, we cheated exactly like that for years. The `Auto` audio backend option only selected `cubeb` every time by default. The shame.
Maide found out that some audio devices incur a huge latency when using `cubeb`, so he decided to {{< gh-hovercard "9039" "perform a latency test" >}} and select either `cubeb` or `SDL` based on the results.
```

[toastUnlimited](https://github.com/lat9nq) then {{< gh-hovercard "9080" "addressed a regression" >}} caused by this change, avoiding a crash.

New Switch firmware versions usually come with their fair share of new services and additions, so in an effort to avoid potential problems in future games, Maide {{< gh-hovercard "9096" "implemented some new parameters" >}} that Nintendo added to their audio core with firmware 15.0.0.

## User interface improvements

> Some users prefer displays closer to the [Golden Ratio](https://en.wikipedia.org/wiki/Golden_ratio), the glorious 16:10 aspect.
> Now, while Switch games are not intended for any ratio different to the standard 16:9, some users are OK with stretching the image to fill their whole display, even if it means some slight distortion.
> This has been increasingly popular with the release of the Steam Deck, which uses a 16:10 display, sporting a 1280x800 resolution.
> There’s also the option of the community coming up with different aspect ratio mods for each game, allowing for proper use of those nice extra pixels.
> german77 took the time to {{< gh-hovercard "9047" "add this new option to the aspect ratio list," >}} which you can find in `Emulation > Configure… > Graphics > Aspect Ratio`.

Reviewer: ```suggestion
Some users prefer displays closer to the [golden ratio](https://en.wikipedia.org/wiki/Golden_ratio), the glorious 16:10 aspect ratio.
Now, while Switch games are not intended for any aspect ratio different to the standard 16:9, some users are fine with stretching the image to fill their whole display, even if it means some slight distortion.
This has been increasingly popular with the release of the Steam Deck, which uses a 16:10 display, sporting a 1280x800 resolution.
There’s also the option of the community coming up with different aspect ratio mods for each game, allowing for proper use of those nice extra pixels.
german77 took the time to {{< gh-hovercard "9047" "add this new option to the aspect ratio list," >}} which you can find in `Emulation > Configure… > Graphics > Aspect Ratio`.
```

{{< imgs
    "./aspectratio.png| I'm personally not a fan of stretching, long live the black bars"
  >}}

> [Docteh](https://github.com/Docteh), Morph, and [lioncash](https://github.com/lioncash) (who diverted some time from the usual, making sure all PRs submitted to yuzu are sane) had their {{< gh-hovercard "9076" "fair share" >}} of fixing {{< gh-hovercard "9079" "spelling mistakes." >}}. 
> Happens to the best of us.

Reviewer: ```suggestion
[Docteh](https://github.com/Docteh), Morph, and [lioncash](https://github.com/lioncash) (who diverted some time from the usual, making sure all PRs submitted to yuzu are sane) had their {{< gh-hovercard "9076" "fair share" >}} of fixing {{< gh-hovercard "9079" "spelling mistakes." >}}.
Happens to the best of us.
```

> In an effort to help reduce user confusion while we rework our compatibility reports, Docteh added a new option to {{< gh-hovercard "9091" "hide the compatibility rating column" >}} from the game list by default.
> Anyone interested in reverting it back, can find it in `Emulation > Configure… > General > UI > Show Compatibility List`.
> Keep in mind reports will not be accurate while we work to implement the new report system.

Reviewer: ```suggestion
In an effort to help reduce user confusion while we rework our compatibility reports, Docteh added a new option to {{< gh-hovercard "9091" "hide the compatibility rating column" >}} from the game list by default.
Anyone interested in reverting it back can find it in `Emulation > Configure… > General > UI > Show Compatibility List`.
Keep in mind, reports will not be accurate while we work to implement the new report system.
```

{{< imgs
    "./compat.png| We're working to solve all those Not Tested reports"
  >}}

Not everyone prefers to have their games listed in English in the game list, many would rather have them in their native language, or at least another option from the official languages the Switch supports.
vonchenplus worked to solve this, and now yuzu {{< gh-hovercard "9115" "prioritizes displaying the games in the user selected language." >}}

{{< single-title-imgs-compare
    "Even if working online forces you to constantly use English, reading in your native language is always the best"
    "./langbug.png"
    "./langfix.png"
>}}

## Hardware section

> This section will be short this month, most stuff is in progress.

Reviewer: ```suggestion
This section will be short this month as most things are in progress!
```

> #### AMD WIP
> 
> As we said previously, we got confirmation from AMD that they are looking into the cause for the crashes at boot now affecting `Xenoblade Chronicles 3`, along with other previously reported fixes.
> 
> Future drivers will be fun to test.
> 
> #### NVIDIA, one step closer
> 
> We’re making some slow progress in the matter of stability problems affecting Maxwell and Pascal cards.
> It’s too early to promise anything, so we’ll keep you updated.
> In the mean time, if you MUST run the latest drivers, use OpenGL.
> 
> Hopefully the next mid-range cards won't cost an entire PC.
> 
> #### Intel ARC, let’s go!
> 
> With games now working on Vulkan thanks to byte[], we only need proud owners of Team Blue cards reporting issues on our GitHub.
> Intel, there are sales outside of Europe, China, and the USA, please start shipping already so your writer can start testing… 
> 
> This goes for you too, Gaben.

Reviewer: ```suggestion
#### AMD, WIP

As we said previously, we got confirmation from AMD that they are looking into the cause for the crashes at boot now affecting `Xenoblade Chronicles 3`, along with other previously reported fixes.

Future drivers will be fun to test.

#### NVIDIA, one step closer

We’re making some slow progress in the matter of stability problems affecting Maxwell and Pascal cards.
It’s too early to promise anything, so we’ll keep you updated.
In the meantime, if you MUST run the latest drivers, use OpenGL.

Hopefully the next mid-range cards won't be the same cost as an entire PC.

#### Intel ARC, let’s go!

With games now working on Vulkan thanks to byte[], we only need proud owners of Team Blue cards reporting issues on our GitHub.
Message to Intel: there are sales outside of Europe, China, and the USA. Please start shipping already so your writer can start testing… 

This goes for you too, Gaben.
```

## Future projects

> With all the regressions caused by the first part of `Project Y.F.C.`, Blinkhawk has started passing us internal testers with the first test builds of some of the changes Part 2 will include.
> As with any first try, there are of course regressions, but some long-standing issues were already gone in those builds, so expect more news in future articles.
> 
> Not forgetting our file system rewrite, `Project Gaia`, Morph implemented {{< gh-hovercard "9082" "support for the incoming save data path structure." >}}
> Work on `Gaia` continues, some roadblocks were hit, slowing progress down, but Morph pushes on.

Reviewer: ```suggestion
With all the regressions caused by the first part of `Project Y.F.C.`, Blinkhawk has started providing internal testers with the first test builds of some of the changes Part 2 will include.
As with any first try, there are of course regressions, but some long-standing issues are already fixed in these builds, so expect more news in future articles.

Not forgetting our file system rewrite, `Project Gaia`, Morph implemented {{< gh-hovercard "9082" "support for the incoming save data path structure." >}}
Work on `Gaia` continues, some roadblocks were hit, slowing progress down, but Morph pushes on.
```

Special thanks to Mysterious Writer B for their big help while your writer is half dead dealing with classes.

{{< imgs
    "./byte.png| He's Vengeance. He's the Night"
  >}}

> That’s all folks! Thank you for your time. See you next month!

Reviewer: ```suggestion
That’s all folks! Thank you for your time and see you next month!
```

Thank you!

## Audio changes

Have you ever noticed how in motherboard BIOS/UEFI settings, the `Auto` option seems to do nothing, and instead simply reflects the first option available in the list?
Well, we cheated exactly like that for years. The `Auto` audio backend option only selected `cubeb` every time by default. The shame.
Maide found out that some audio devices incur a huge latency when using `cubeb`, so he decided to {{< gh-hovercard "9039" "perform a latency test" >}} and select either `cubeb` or `SDL` based on the results.

[toastUnlimited](https://github.com/lat9nq) then {{< gh-hovercard "9080" "addressed a regression" >}} caused by this change, avoiding a crash.

New Switch firmware versions usually come with their fair share of new services and additions, so in an effort to avoid potential problems in future games, Maide {{< gh-hovercard "9096" "implemented some new parameters" >}} that Nintendo added to their audio core with firmware 15.0.0.

## User interface improvements

Some users prefer displays closer to the [golden ratio](https://en.wikipedia.org/wiki/Golden_ratio), the glorious 16:10 aspect ratio.

> Now, while Switch games are not intended for any aspect ratio different to the standard 16:9, some users are fine with stretching the image to fill their whole display, even if it means some slight distortion.

Reviewer: ```suggestion
While Switch games are not intended to be played outside the standard 16:9 aspect ratio, some users are fine with stretching the image to fill their whole display, even if it means some slight distortion.
```

This has been increasingly popular with the release of the Steam Deck, which uses a 16:10 display, sporting a 1280x800 resolution.
There’s also the option of the community coming up with different aspect ratio mods for each game, allowing for proper use of those nice extra pixels.
german77 took the time to {{< gh-hovercard "9047" "add this new option to the aspect ratio list," >}} which you can find in `Emulation > Configure… > Graphics > Aspect Ratio`.

{{< imgs
    "./aspectratio.png| I'm personally not a fan of stretching, long live the black bars"
  >}}

> [Docteh](https://github.com/Docteh), Morph, and [lioncash](https://github.com/lioncash) (who diverted some time from the usual, making sure all PRs submitted to yuzu are sane) had their {{< gh-hovercard "9076" "fair share" >}} of fixing {{< gh-hovercard "9079" "spelling mistakes." >}}.

Reviewer: ```suggestion
[Docteh](https://github.com/Docteh), Morph, and [lioncash](https://github.com/lioncash) (who diverted some time from the usual, making sure all PRs submitted to yuzu are sane) had their {{< gh-hovercard "9076" "fair share" >}} of fixing {{< gh-hovercard "9079" "spelling mistakes." >}}
```
